db/system_keyspace.cc

```diff
@@ -3398,11 +3398,11 @@ mutation system_keyspace::make_group0_history_state_id_mutation(
         using namespace std::chrono;
         assert(*gc_older_than >= gc_clock::duration{0});
 
-        auto ts_millis = duration_cast<milliseconds>(microseconds{ts});
-        auto gc_older_than_millis = duration_cast<milliseconds>(*gc_older_than);
-        assert(gc_older_than_millis < ts_millis);
+        auto ts_micros = microseconds{ts};
+        auto gc_older_than_micros = duration_cast<microseconds>(*gc_older_than);
+        assert(gc_older_than_micros < ts_micros);
 
-        auto tomb_upper_bound = utils::UUID_gen::min_time_UUID(ts_millis - gc_older_than_millis);
+        auto tomb_upper_bound = utils::UUID_gen::min_time_UUID(ts_micros - gc_older_than_micros);
```

Comment:
I don't understand one thing. You say that:

> The consecutive state IDs used in group 0 mutations are guaranteed to be strictly monotonic at
microsecond resolution 

Then why is generate_group0_state_id() using get_random_time_UUID_from_micros() to generate timestamps:

```
    return utils::UUID_gen::get_random_time_UUID_from_micros(std::chrono::microseconds{ts});
```

If ts was guaranteed to be monotonic, why not use `min_time_UUID()` like here.

Reply:
Hm, thinking about it more -- I lied.
They will be strictly monotonic in this particular test/scenario - when there's one node sending the requests.
But in general two nodes performing a group 0 operation could pick the same timestamp, hence we need timeUUIDs to resolve the collision, so that checking for `group0_concurrent_modification` works correctly.